### ISSUE TITLE
BenchmarkConfig

### ISSUE THREAD

**Issue by AndreyAkinshin** [PerfDotNet/BenchmarkDotNet#82]
I suggest to make a big refactoring. New BenchmarkDotNet will look like follows:
```cs
[BenchmarkConfig(typeof(MyBenchmarkConfig)]
public class MyBenchmarkClass
{
    [Benchmark]
    public void MyBenchmarkMethod()
    {
        // ...
    }
}
public class MyBenchmarkConfig : BenchmarkConfig
{
    public override IEnumerable<BenchmarkTask> GetTasks()
    {
         yield return new BenchmarkTask
         { 
             Platform = Platform.X86, 
             Jitter = Jitter.LegacyJit
         };
    }
    public override IEnumerable<BenchmarkPlugin> GetPlugins()
    {
        yield return new MyBenchmarkPlugin();
    }
    // ...
}
```
Advantages:
* Nice tasks. Current BenchmarkTasks look awful because of big amount of arguments in the constructor, we can rewrite it in nice way.
* We can implement a lot of features like [tags](https://github.com/PerfDotNet/BenchmarkDotNet/issues/74) because we can define additional config stuff in the method instead of attributes.
* We can use static instances like `BenchmarkStatResultExtender.StdDev`
* We can write some predefined common configs (e.g. a config that contains LegacyJit-x86, LegacyJit-x64, RyuJIT-x64)
* We can easily override [BenchmarkConfigAttribute] by custom config in the `BenchmarkRunner.Run` call or specify config for a 3rdParty class that doesn't have such attribute.

@mattwarren, what do you think?

**Comment by mattwarren** [comment]
Looks like a nice idea, yeah `[BenchmarkTask(..)]` has got a bit unwieldy.

I definitely agree that we need to do this, so that we aren't expecting users to write their own `MyBenchmarkConfig` class just to run a simple benchmark:

> We can write some predefined common configs (e.g. a config that contains LegacyJit-x86, LegacyJit-x64, RyuJIT-x64)

One question, is it a good idea (or is it even possible) to keep the old way around for backwards compatability? Or is now a good time to break it? It might be a bit much to expect existing users to fix all their benchmarks if they upgrade.

Just 2 "dis-advantages" I can think of:
- you can no longer see the options at the top of the benchmark, they'll be hidden away somewhere else, i.e. `LegacyJit-x86` or `MyBenchmarkConfig`
- sharing benchmarks in a gist is trickier if you need to also paste the custom `MyBenchmarkConfig` class

But these aren't reasons not to do it, I just wanted to bring them up.

**Comment by AndreyAkinshin** [comment]
> Or is now a good time to break it?

Here is the plan: I will make v0.9 with support old way (but we can mark `BenchmakTask` as an obsolete attribute); in v0.10 we will drop attributes (we can do it in a few months)

> you can no longer see the options at the top of the benchmark

I don't think that it is a problem. When you look at the source code, you usually want to look only at the source code. When you look at the benchmark results, options are obvious. If you really want to see config, you can scroll down. 

> sharing benchmarks in a gist is trickier if you need to also paste the custom MyBenchmarkConfig class

If you want to use custom config, you can put it in the same file below your benchmark class.

**Comment by AndreyAkinshin** [comment]
Implemented in `v0.9.0`.
